icu4j/main/classes/core/src/com/ibm/icu/text/MeasureFormat.java

```diff
@@ -861,116 +859,77 @@ private static DateFormat loadNumericDurationFormat(ICUResourceBundle r, String
-            Date duration,
-            DateFormat formatter,
-            DateFormat.Field smallestField,
-            Number smallestAmount,
-            Appendable appendTo) {
-        // Format the smallest amount ahead of time.
-        String smallestAmountFormatted;
-
-        // Format the smallest amount using this object's number format, but keep track
-        // of the integer portion of this formatted amount. We have to replace just the
-        // integer part with the corresponding value from formatting the date. Otherwise
-        // when formatting 0 minutes 9 seconds, we may get "00:9" instead of "00:09"
-        FieldPosition intFieldPosition = new FieldPosition(NumberFormat.INTEGER_FIELD);
-        FormattedNumber result = getNumberFormatter().format(smallestAmount);
-        result.nextFieldPosition(intFieldPosition);
-        smallestAmountFormatted = result.toString();
-        // Give up if there is no integer field.
-        if (intFieldPosition.getBeginIndex() == 0 && intFieldPosition.getEndIndex() == 0) {
-            throw new IllegalStateException();
-        }
+        // We can create it on demand, but all of the patterns (right now) have mm and ss.
+        // So unless it is hours only we will need a 0-padded 2 digits formatter.
+        LocalizedNumberFormatter numberFormatter2 = numberFormatter.integerWidth(IntegerWidth.zeroFillTo(2));
+        FormattedStringBuilder fsb = new FormattedStringBuilder();
+
+        boolean protect = false;
+        for (int i = 0; i < pattern.length(); i++) {
+            Character c = pattern.charAt(i);
+
+            // Also set the proper field in this switch
+            // We don't use DateFormat.Field because this is not a date / time, is a duration.
+            Number value = 0;
+            switch (c) {
+                case 'H': value = hms[0]; break;
+                case 'm': value = hms[1]; break;
+                case 's': value = hms[2]; break;
+            }
 
-        // Format our duration as a date, but keep track of where the smallest field is
-        // so that we can use it to replace the integer portion of the smallest value.
-        // #13606: DateFormat is not thread-safe, but MeasureFormat advertises itself as thread-safe.
-        FieldPosition smallestFieldPosition = new FieldPosition(smallestField);
-        String draft;
-        synchronized (formatter) {
-            draft = formatter.format(duration, new StringBuffer(), smallestFieldPosition).toString();
+            // There is not enough info to add Field(s) for the unit because all we have are plain
+            // text patterns. For example in "21:51" there is no text for something like "hour",
+            // while in something like "21h51" there is ("h"). But we can't really tell...
+            switch (c) {
+                case 'H':
+                case 'm':
+                case 's':
+                    if (protect) {
+                        fsb.append(c.toString(), null);
```

Comment:
Please use `fsb.appendCodePoint(c, null)`

Reply:
This would damage strings with surrogates.
Java string is UTF-16, so a char (or Character) is not a code point, is a code unit.

Reply:
> This would damage strings with surrogates.

The implementation checks if it needs to allocate 1 or 2 code units.  In C++, it checks with U16_LENGTH, and in Java with Character.charCount.

C++ U16_LENGTH is defined as:

    #define U16_LENGTH(c) ((uint32_t)(c)<=0xffff ? 1 : 2)

Java Character.charCount is defined as:

> If the specified character is equal to or greater than 0x10000, then the method returns 2. Otherwise, the method returns 1.

Therefore I think the method works even if you are dealing with a surrogate code unit.

I filed a ticket to clarify: https://unicode-org.atlassian.net/browse/ICU-20666

FormattedStringBuilder is an internal API that isn't perfect but we continually improve on it.

> The Java StringBuilder has an append(char), but FormattedStringBuilder does not.

It was built with the idea that you operate in code points.

Reply:
Another way to put it is, a code unit (including a surrogate code unit) can be treated as if it were a code point.

Reply:
Java style is to loop over code units.  ICU style is to loop over code points (but to index into strings using code unit offsets).  That's why the ICU internal API has an appendCodePoint, but the Java API has an append(char).

If you want to re-write your loop ICU-style, you could do:

```java
for (int cp, i = 0; i < pattern.length(); i += Character.charCount(cp)) {
  cp = Character.codePointAt(pattern, i);
  // ...
}
```

The alternative is to use `appendCodePoint(c, null)` despite the misnomer.  Converting to a string is not efficient and isn't a good option.

Reply:
Done